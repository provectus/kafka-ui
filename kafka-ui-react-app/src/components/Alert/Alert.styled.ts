import styled from 'styled-components';
import { Colors } from 'theme/theme';

type Type = 'error' | 'success' | 'warning' | 'info';

function getColor(type: Type) {
  switch (type) {
    case 'error':
      return Colors.red[10];
    case 'success':
      return Colors.green[10];
    case 'warning':
      return Colors.yellow[10];
    default:
      return Colors.neutral[10];
  }
}

export const Wrapper = styled.div<{ $type: Type }>`
  background-color: ${({ $type }) => getColor($type)};
  width: 400px;
  min-height: 64px;
  border-radius: 8px;
  padding: 12px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  filter: drop-shadow(0px 4px 16px rgba(0, 0, 0, 0.1));
  margin-top: 10px;
  line-height: 20px;
`;

export const Title = styled.div`
  font-weight: 500;
  font-size: 14px;
`;

export const Message = styled.p`
  font-weight: normal;
  font-size: 14px;
  margin: 3px 0;
`;
