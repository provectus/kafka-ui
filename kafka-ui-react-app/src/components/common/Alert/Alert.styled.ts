import styled from 'styled-components';
import { ToastTypes } from 'lib/errorHandling';

export const Alert = styled.div<{ $type: ToastTypes }>`
  background-color: ${({ $type, theme }) => theme.alert.color[$type]};
  width: 500px;
  min-height: 64px;
  border-radius: 8px;
  padding: 12px;
  display: flex;
  justify-content: space-between;
  align-items: baseline;
  filter: drop-shadow(0px 4px 16px ${({ theme }) => theme.alert.shadow});
  margin-top: 10px;
  line-height: 20px;
`;

export const Title = styled.div`
  font-weight: 500;
  font-size: 14px;
`;

export const Message = styled.div`
  font-weight: normal;
  font-size: 14px;
  margin: 3px 0;

  ol,
  ul {
    padding-left: 25px;
    list-style: auto;
  }
`;
