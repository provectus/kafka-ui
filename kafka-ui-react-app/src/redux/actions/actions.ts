import { createAsyncAction } from 'typesafe-actions';
import ActionType from 'redux/actionType';
import {
  Broker,
  BrokerMetrics,
  Cluster,
  Topic,
  TopicConfig,
  TopicDetails,
  TopicMessage,
  TopicName,
  ConsumerGroup,
  ConsumerGroupDetails,
  ConsumerGroupID,
} from 'redux/interfaces';

export const fetchBrokersAction = createAsyncAction(
  ActionType.GET_BROKERS__REQUEST,
  ActionType.GET_BROKERS__SUCCESS,
  ActionType.GET_BROKERS__FAILURE
)<undefined, Broker[], undefined>();

export const fetchBrokerMetricsAction = createAsyncAction(
  ActionType.GET_BROKER_METRICS__REQUEST,
  ActionType.GET_BROKER_METRICS__SUCCESS,
  ActionType.GET_BROKER_METRICS__FAILURE
)<undefined, BrokerMetrics, undefined>();

export const fetchClusterListAction = createAsyncAction(
  ActionType.GET_CLUSTERS__REQUEST,
  ActionType.GET_CLUSTERS__SUCCESS,
  ActionType.GET_CLUSTERS__FAILURE
)<undefined, Cluster[], undefined>();

export const fetchTopicListAction = createAsyncAction(
  ActionType.GET_TOPICS__REQUEST,
  ActionType.GET_TOPICS__SUCCESS,
  ActionType.GET_TOPICS__FAILURE
)<undefined, Topic[], undefined>();

export const fetchTopicMessagesAction = createAsyncAction(
  ActionType.GET_TOPIC_MESSAGES__REQUEST,
  ActionType.GET_TOPIC_MESSAGES__SUCCESS,
  ActionType.GET_TOPIC_MESSAGES__FAILURE
)<undefined, TopicMessage[], undefined>();

export const fetchTopicDetailsAction = createAsyncAction(
  ActionType.GET_TOPIC_DETAILS__REQUEST,
  ActionType.GET_TOPIC_DETAILS__SUCCESS,
  ActionType.GET_TOPIC_DETAILS__FAILURE
)<undefined, { topicName: TopicName; details: TopicDetails }, undefined>();

export const fetchTopicConfigAction = createAsyncAction(
  ActionType.GET_TOPIC_CONFIG__REQUEST,
  ActionType.GET_TOPIC_CONFIG__SUCCESS,
  ActionType.GET_TOPIC_CONFIG__FAILURE
)<undefined, { topicName: TopicName; config: TopicConfig[] }, undefined>();

export const createTopicAction = createAsyncAction(
  ActionType.POST_TOPIC__REQUEST,
  ActionType.POST_TOPIC__SUCCESS,
  ActionType.POST_TOPIC__FAILURE
)<undefined, Topic, undefined>();

export const updateTopicAction = createAsyncAction(
  ActionType.PATCH_TOPIC__REQUEST,
  ActionType.PATCH_TOPIC__SUCCESS,
  ActionType.PATCH_TOPIC__FAILURE
)<undefined, Topic, undefined>();

export const fetchConsumerGroupsAction = createAsyncAction(
  ActionType.GET_CONSUMER_GROUPS__REQUEST,
  ActionType.GET_CONSUMER_GROUPS__SUCCESS,
  ActionType.GET_CONSUMER_GROUPS__FAILURE
)<undefined, ConsumerGroup[], undefined>();

export const fetchConsumerGroupDetailsAction = createAsyncAction(
  ActionType.GET_CONSUMER_GROUP_DETAILS__REQUEST,
  ActionType.GET_CONSUMER_GROUP_DETAILS__SUCCESS,
  ActionType.GET_CONSUMER_GROUP_DETAILS__FAILURE
)<
  undefined,
  { consumerGroupID: ConsumerGroupID; details: ConsumerGroupDetails },
  undefined
>();
