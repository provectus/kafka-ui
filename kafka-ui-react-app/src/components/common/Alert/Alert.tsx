import React from 'react';
import CloseIcon from 'components/common/Icons/CloseIcon';
import IconButtonWrapper from 'components/common/Icons/IconButtonWrapper';
import { ToastTypes } from 'lib/errorHandling';

import * as S from './Alert.styled';

export interface AlertProps {
  title: string;
  type: ToastTypes;
  message: React.ReactNode;
  onDissmiss(): void;
}

const Alert: React.FC<AlertProps> = ({ title, type, message, onDissmiss }) => (
  <S.Alert $type={type} role="alert">
    <div>
      <S.Title role="heading">{title}</S.Title>
      <S.Message role="contentinfo">{message}</S.Message>
    </div>
    <IconButtonWrapper role="button" onClick={onDissmiss}>
      <CloseIcon />
    </IconButtonWrapper>
  </S.Alert>
);

export default Alert;
