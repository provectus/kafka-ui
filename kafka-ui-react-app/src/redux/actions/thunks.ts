import {
  ApiClustersApi,
  Configuration,
  Cluster,
  Topic,
  TopicFormData,
  TopicConfig,
  NewSchemaSubject,
  SchemaSubject,
} from 'generated-sources';
import {
  ConsumerGroupID,
  PromiseThunkResult,
  ClusterName,
  BrokerId,
  TopicName,
  TopicMessageQueryParams,
  TopicFormFormattedParams,
  TopicFormDataRaw,
  SchemaName,
} from 'redux/interfaces';

import { BASE_PARAMS } from 'lib/constants';
import * as actions from './actions';

const apiClientConf = new Configuration(BASE_PARAMS);
export const apiClient = new ApiClustersApi(apiClientConf);

export const fetchClustersList = (): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.fetchClusterListAction.request());
  try {
    const clusters: Cluster[] = await apiClient.getClusters();
    dispatch(actions.fetchClusterListAction.success(clusters));
  } catch (e) {
    dispatch(actions.fetchClusterListAction.failure());
  }
};

export const fetchClusterStats = (
  clusterName: ClusterName
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.fetchClusterStatsAction.request());
  try {
    const payload = await apiClient.getClusterStats({ clusterName });
    dispatch(actions.fetchClusterStatsAction.success(payload));
  } catch (e) {
    dispatch(actions.fetchClusterStatsAction.failure());
  }
};

export const fetchClusterMetrics = (
  clusterName: ClusterName
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.fetchClusterMetricsAction.request());
  try {
    const payload = await apiClient.getClusterMetrics({ clusterName });
    dispatch(actions.fetchClusterMetricsAction.success(payload));
  } catch (e) {
    dispatch(actions.fetchClusterMetricsAction.failure());
  }
};

export const fetchBrokers = (
  clusterName: ClusterName
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.fetchBrokersAction.request());
  try {
    const payload = await apiClient.getBrokers({ clusterName });
    dispatch(actions.fetchBrokersAction.success(payload));
  } catch (e) {
    dispatch(actions.fetchBrokersAction.failure());
  }
};

export const fetchBrokerMetrics = (
  clusterName: ClusterName,
  brokerId: BrokerId
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.fetchBrokerMetricsAction.request());
  try {
    const payload = await apiClient.getBrokersMetrics({
      clusterName,
      id: brokerId,
    });
    dispatch(actions.fetchBrokerMetricsAction.success(payload));
  } catch (e) {
    dispatch(actions.fetchBrokerMetricsAction.failure());
  }
};

export const fetchTopicsList = (
  clusterName: ClusterName
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.fetchTopicsListAction.request());
  try {
    const topics = await apiClient.getTopics({ clusterName });
    dispatch(actions.fetchTopicsListAction.success(topics));
  } catch (e) {
    dispatch(actions.fetchTopicsListAction.failure());
  }
};

export const fetchTopicMessages = (
  clusterName: ClusterName,
  topicName: TopicName,
  queryParams: Partial<TopicMessageQueryParams>
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.fetchTopicMessagesAction.request());
  try {
    const messages = await apiClient.getTopicMessages({
      clusterName,
      topicName,
      ...queryParams,
    });
    dispatch(actions.fetchTopicMessagesAction.success(messages));
  } catch (e) {
    dispatch(actions.fetchTopicMessagesAction.failure());
  }
};

export const fetchTopicDetails = (
  clusterName: ClusterName,
  topicName: TopicName
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.fetchTopicDetailsAction.request());
  try {
    const topicDetails = await apiClient.getTopicDetails({
      clusterName,
      topicName,
    });
    dispatch(
      actions.fetchTopicDetailsAction.success({
        topicName,
        details: topicDetails,
      })
    );
  } catch (e) {
    dispatch(actions.fetchTopicDetailsAction.failure());
  }
};

export const fetchTopicConfig = (
  clusterName: ClusterName,
  topicName: TopicName
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.fetchTopicConfigAction.request());
  try {
    const config = await apiClient.getTopicConfigs({ clusterName, topicName });
    dispatch(actions.fetchTopicConfigAction.success({ topicName, config }));
  } catch (e) {
    dispatch(actions.fetchTopicConfigAction.failure());
  }
};

const formatTopicFormData = (form: TopicFormDataRaw): TopicFormData => {
  const {
    name,
    partitions,
    replicationFactor,
    cleanupPolicy,
    retentionBytes,
    retentionMs,
    maxMessageBytes,
    minInSyncReplicas,
    customParams,
  } = form;

  return {
    name,
    partitions,
    replicationFactor,
    configs: {
      'cleanup.policy': cleanupPolicy,
      'retention.ms': retentionMs,
      'retention.bytes': retentionBytes,
      'max.message.bytes': maxMessageBytes,
      'min.insync.replicas': minInSyncReplicas,
      ...Object.values(customParams || {}).reduce(
        (result: TopicFormFormattedParams, customParam: TopicConfig) => {
          return {
            ...result,
            [customParam.name]: customParam.value,
          };
        },
        {}
      ),
    },
  };
};

export const createTopic = (
  clusterName: ClusterName,
  form: TopicFormDataRaw
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.createTopicAction.request());
  try {
    const topic: Topic = await apiClient.createTopic({
      clusterName,
      topicFormData: formatTopicFormData(form),
    });
    dispatch(actions.createTopicAction.success(topic));
  } catch (e) {
    dispatch(actions.createTopicAction.failure());
  }
};

export const updateTopic = (
  clusterName: ClusterName,
  form: TopicFormDataRaw
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.updateTopicAction.request());
  try {
    const topic: Topic = await apiClient.updateTopic({
      clusterName,
      topicName: form.name,
      topicFormData: formatTopicFormData(form),
    });
    dispatch(actions.updateTopicAction.success(topic));
  } catch (e) {
    dispatch(actions.updateTopicAction.failure());
  }
};

export const fetchConsumerGroupsList = (
  clusterName: ClusterName
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.fetchConsumerGroupsAction.request());
  try {
    const consumerGroups = await apiClient.getConsumerGroups({ clusterName });
    dispatch(actions.fetchConsumerGroupsAction.success(consumerGroups));
  } catch (e) {
    dispatch(actions.fetchConsumerGroupsAction.failure());
  }
};

export const fetchConsumerGroupDetails = (
  clusterName: ClusterName,
  consumerGroupID: ConsumerGroupID
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.fetchConsumerGroupDetailsAction.request());
  try {
    const consumerGroupDetails = await apiClient.getConsumerGroup({
      clusterName,
      id: consumerGroupID,
    });
    dispatch(
      actions.fetchConsumerGroupDetailsAction.success({
        consumerGroupID,
        details: consumerGroupDetails,
      })
    );
  } catch (e) {
    dispatch(actions.fetchConsumerGroupDetailsAction.failure());
  }
};

export const fetchSchemasByClusterName = (
  clusterName: ClusterName
): PromiseThunkResult<void> => async (dispatch) => {
  dispatch(actions.fetchSchemasByClusterNameAction.request());
  try {
    const schemas = await apiClient.getSchemas({ clusterName });
    dispatch(actions.fetchSchemasByClusterNameAction.success(schemas));
  } catch (e) {
    dispatch(actions.fetchSchemasByClusterNameAction.failure());
  }
};

export const fetchSchemaVersions = (
  clusterName: ClusterName,
  subject: SchemaName
): PromiseThunkResult<void> => async (dispatch) => {
  if (!subject) return;
  dispatch(actions.fetchSchemaVersionsAction.request());
  try {
    const versions = await apiClient.getAllVersionsBySubject({
      clusterName,
      subject,
    });
    dispatch(actions.fetchSchemaVersionsAction.success(versions));
  } catch (e) {
    dispatch(actions.fetchSchemaVersionsAction.failure());
  }
};

export const createSchema = (
  clusterName: ClusterName,
  newSchemaSubject: NewSchemaSubject
): PromiseThunkResult => async (dispatch) => {
  dispatch(actions.createSchemaAction.request());
  try {
    const schema: SchemaSubject = await apiClient.createNewSchema({
      clusterName,
      newSchemaSubject,
    });
    dispatch(actions.createSchemaAction.success(schema));
  } catch (e) {
    dispatch(actions.createSchemaAction.failure());
    throw e;
  }
};
