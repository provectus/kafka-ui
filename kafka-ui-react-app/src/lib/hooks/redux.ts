import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { AppDispatch, RootState } from 'redux/interfaces';

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
