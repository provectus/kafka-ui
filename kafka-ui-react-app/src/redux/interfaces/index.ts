import rootReducer from 'redux/reducers';
import { store } from 'redux/store';

export * from './topic';
export * from './cluster';
export * from './consumerGroup';
export * from './schema';
export * from './loader';

export type RootState = ReturnType<typeof rootReducer>;
export type AppDispatch = typeof store.dispatch;
