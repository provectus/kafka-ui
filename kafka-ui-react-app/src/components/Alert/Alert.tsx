import React from 'react';
import cx from 'classnames';
import { useDispatch } from 'react-redux';
import { dismissAlert } from 'redux/actions';
import { Alert as AlertProps } from 'redux/interfaces';

const Alert: React.FC<AlertProps> = ({
  id,
  type,
  title,
  message,
  response,
}) => {
  const classNames = React.useMemo(
    () =>
      cx('notification', {
        'is-danger': type === 'error',
        'is-success': type === 'success',
        'is-info': type === 'info',
        'is-warning': type === 'warning',
      }),
    [type]
  );
  const dispatch = useDispatch();
  const dismiss = React.useCallback(() => {
    dispatch(dismissAlert(id));
  }, []);

  return (
    <div className={classNames}>
      <button className="delete" type="button" onClick={dismiss}>
        x
      </button>
      <div>
        <h6 className="title is-6">{title}</h6>
        <p className="subtitle is-6">{message}</p>
        {response && (
          <div className="is-flex">
            <div className="mr-3">{response.status}</div>
            <div>{response.body?.message || response.statusText}</div>
          </div>
        )}
      </div>
    </div>
  );
};

export default Alert;
