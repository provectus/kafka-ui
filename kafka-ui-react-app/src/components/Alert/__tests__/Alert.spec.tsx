import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Alert as AlertProps } from 'redux/interfaces';
import * as actions from 'redux/actions/actions';
import Alert from 'components/Alert/Alert';

const id = 'test-id';
const title = 'My Alert Title';
const message = 'My Alert Message';
const statusCode = 123;
const serverSideMessage = 'Server Side Message';
const httpStatusText = 'My Status Text';
const dismiss = jest.fn();

describe('Alert', () => {
  const setupComponent = (props: Partial<AlertProps> = {}) =>
    render(
      <Alert
        id={id}
        type="error"
        title={title}
        message={message}
        createdAt={1234567}
        {...props}
      />
    );
  it('renders with initial props', () => {
    setupComponent();
    expect(screen.getByRole('heading')).toHaveTextContent(title);
    expect(screen.getByRole('contentinfo')).toHaveTextContent(message);
    expect(screen.getByRole('button')).toBeInTheDocument();
  });
  it('renders alert with server side message', () => {
    setupComponent({
      type: 'info',
      response: {
        status: statusCode,
        statusText: 'My Status Text',
        body: {
          message: serverSideMessage,
        },
      },
    });
    expect(screen.getAllByRole('contentinfo').length).toEqual(2);
    expect(screen.getByText(message)).toBeInTheDocument();
    expect(
      screen.getByText(`${statusCode} ${serverSideMessage}`)
    ).toBeInTheDocument();
  });
  it('renders alert with http status text', () => {
    setupComponent({
      type: 'warning',
      response: {
        status: statusCode,
        statusText: httpStatusText,
        body: {},
      },
    });
    expect(screen.getAllByRole('contentinfo').length).toEqual(2);
    expect(screen.getByText(message)).toBeInTheDocument();
    expect(
      screen.getByText(`${statusCode} ${httpStatusText}`)
    ).toBeInTheDocument();
  });
  it('handles dismiss callback', () => {
    jest.spyOn(actions, 'dismissAlert').mockImplementation(dismiss);
    setupComponent();
    userEvent.click(screen.getByRole('button'));
    expect(dismiss).toHaveBeenCalledWith(id);
  });
});
