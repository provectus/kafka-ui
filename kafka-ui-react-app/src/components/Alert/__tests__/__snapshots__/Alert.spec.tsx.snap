// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert matches snapshot 1`] = `
<Alert
  createdAt={1234567}
  id="test-id"
  message="My Alert Message"
  title="My Alert Title"
  type="error"
>
  <div
    className="notification is-danger"
  >
    <button
      className="delete"
      onClick={[Function]}
      type="button"
    >
      x
    </button>
    <div>
      <h6
        className="title is-6"
      >
        My Alert Title
      </h6>
      <p
        className="subtitle is-6"
      >
        My Alert Message
      </p>
    </div>
  </div>
</Alert>
`;
